lib/eth/contract.rb

```diff
@@ -0,0 +1,301 @@
+require 'forwardable'
+
+module Eth
+  class Contract
+    attr_reader :address
+    attr_accessor :key
+    attr_accessor :gas_limit, :gas_price, :max_fee_per_gas, :max_priority_fee_per_gas, :nonce
+    attr_accessor :code, :name, :abi, :class_object, :sender, :deployment, :client
+    attr_accessor :events, :functions, :constructor_inputs
+    attr_accessor :call_raw_proxy, :call_proxy, :transact_proxy, :transact_and_wait_proxy
+    attr_accessor :new_filter_proxy, :get_filter_logs_proxy, :get_filter_change_proxy
+
+    def initialize(name, code, abi, client)
+      @name = name
+      @code = code
+      @abi = abi
+      @client = client
+      @constructor_inputs, @functions, @events = Eth::Contract::Abi.parse_abi(abi)
+      @encoder = Eth::Contract::Encoder.new
+      @formatter = Eth::Contract::Formatter.new
+      @decoder = Eth::Contract::Decoder.new
+      @sender = client.default_account
+      @gas_limit = @client.gas_limit
+      @gas_price = @client.max_fee_per_gas
+      @max_fee_per_gas = @client.max_fee_per_gas
+      @max_priority_fee_per_gas = @client.max_priority_fee_per_gas
+    end
+
+    # Creates a contract wrapper.
+    def self.create(file: nil, client: nil, code: nil, abi: nil, address: nil, name: nil, contract_index: nil)
+      contract = nil
+      client = Eth::Client.create "/tmp/geth.ipc"
+      if File.exist?(file)
+        contracts = Eth::Contract::Initializer.new(file, client).build_all
+        raise "No contracts compiled" if contracts.empty?
+        if contract_index
+          contract = contracts[contract_index].class_object.new
+        else
+          contract = contracts.first.class_object.new
+        end
+      else
+        abi = abi.is_a?(String) ? JSON.parse(abi) : abi.map(&:deep_stringify_keys)
+        contract = Eth::Contract.new(name, code, abi, client)
+        contract.build
+        contract = contract.class_object.new
+      end
+      contract.address = address
+      contract
+    end
+
+    def address=(addr)
+      @address = addr
```

Comment:
address should always be an Eth::Address for security reasons.

Reply:
I now use Eth::Address 📝